File: trunk/FftGuitarTuner/SoundFrequencyInfoSource.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using SoundCapture;
using SoundAnalysis;

namespace FftGuitarTuner
{
    class SoundFrequencyInfoSource : FrequencyInfoSource
    {
        SoundCaptureDevice device;
        Adapter adapter;

        internal SoundFrequencyInfoSource(SoundCaptureDevice device)
        {
            this.device = device;
        }

        public override void Listen()
        {
            adapter = new Adapter(this, device);
            adapter.Start();
        }

        public override void Stop()
        {
            adapter.Stop();
        }

        class Adapter : SoundCaptureBase
        {
            SoundFrequencyInfoSource owner;

            const double MinFreq = 60;
            const double MaxFreq = 1300;

            internal Adapter(SoundFrequencyInfoSource owner, SoundCaptureDevice device)
                : base(device)
            {
                this.owner = owner;
            }

            protected override void ProcessData(short[] data)
            {
                double[] x = new double[data.Length];
                for (int i = 0; i < x.Length; i++)
                {
                    x[i] = data[i] / 32768.0;
                }

                double freq = FrequencyUtils.FindFundamentalFrequency(x, SampleRate, MinFreq, MaxFreq);
                owner.OnFrequencyDetected(new FrequencyDetectedEventArgs(freq));
            }
        }
    }
}

File: vb.net/FftGuitarTuner/SoundFrequencyInfoSource.vb
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports SoundCapture
Imports SoundAnalysis

Class SoundFrequencyInfoSource
	Inherits FrequencyInfoSource
	Private device As SoundCaptureDevice
	Private adapter As _Adapter

	Friend Sub New(device As SoundCaptureDevice)
		Me.device = device
	End Sub

	Public Overrides Sub Listen()
		adapter = New _Adapter(Me, device)
		adapter.Start()
	End Sub

	Public Overrides Sub [Stop]()
		adapter.[Stop]()
	End Sub

	Private Class _Adapter
		Inherits SoundCaptureBase
		Private owner As SoundFrequencyInfoSource

		Const MinFreq As Double = 60
		Const MaxFreq As Double = 1300

		Friend Sub New(owner As SoundFrequencyInfoSource, device As SoundCaptureDevice)
			MyBase.New(device)
			Me.owner = owner
		End Sub

		Protected Overrides Sub ProcessData(data As Short())
			Dim x As Double() = New Double(data.Length - 1) {}
			For i As Integer = 0 To x.Length - 1
				x(i) = data(i) / 32768.0
			Next

			Dim freq As Double = FrequencyUtils.FindFundamentalFrequency(x, SampleRate, MinFreq, MaxFreq)
			owner.OnFrequencyDetected(New FrequencyDetectedEventArgs(freq))
		End Sub
	End Class
End Class
